controller.go

```diff
@@ -952,29 +953,50 @@ func parseAllocateConstraintsResponse(source interface{}, machine *machine) (Con
 	if storageMatches, found := constraintsMap["storage"]; found {
 		matches := convertConstraintMatches(storageMatches)
 		for label, ids := range matches {
-			blockDevices := make([]BlockDevice, len(ids))
-			for index, id := range ids {
-				blockDevice := machine.BlockDevice(id)
-				if blockDevice == nil {
-					return empty, NewDeserializationError("constraint match storage %q: %d does not match a block device for the machine", label, id)
+			storageDevices := make([]StorageDevice, len(ids))
+			for index, storageId := range ids {
+				if id, ok := storageId.(float64); ok {
+					// Links to a block device.
+					blockDevice := machine.BlockDevice(int(id))
+					if blockDevice == nil {
+						return empty, NewDeserializationError("constraint match storage %q: %d does not match a block device for the machine", label, int(id))
+					}
+					storageDevices[index] = blockDevice
+				} else if id, ok := storageId.(string); ok {
+					// Should link to a partition.
+					const partPrefix = "partition:"
+					if strings.HasPrefix(id, partPrefix) {
+						partId, err := strconv.Atoi(id[len(partPrefix):])
+						if err != nil {
+							return empty, NewDeserializationError("constraint match storage %q: %s cannot convert to int.", label, id[len(partPrefix):])
+						}
+						partition := machine.Partition(partId)
+						if partition == nil {
+							return empty, NewDeserializationError("constraint match storage %q: %d does not match a partition for the machine", label, partId)
+						}
+						storageDevices[index] = partition
+					} else {
+						return empty, NewDeserializationError("constraint match storage %q: %s is not prefixed with partition", label, id)
+					}
+				} else {
+					return empty, NewDeserializationError("constraint match storage %q: %v is not an int or string", label, storageId)
 				}
-				blockDevices[index] = blockDevice
 			}
-			result.Storage[label] = blockDevices
+			result.Storage[label] = storageDevices
 		}
 	}
 	return result, nil
 }
 
-func convertConstraintMatches(source interface{}) map[string][]int {
+func convertConstraintMatches(source interface{}) map[string][]interface{} {
```

Comment:
As stated above, maybe 2 versions of this to return the expected typed map and error if the values don't conform